### ISSUE TITLE
High VIRT memory 

### ISSUE THREAD

**Issue by RetroZelda** [slskd/slskd#1112]
I am running slskd on a headless debian server(Trixie; 6.6.13 kernel) and I noticed a really high virtual memory usage(260GB).  The music collection i have shared is about 920GB(7238directories and 38563 files) and the folder where downloads go is empty.   Im relativly new to slskd/soulseek so I'm wondering if this is expected, or if I have something configured incorrectly. Thanks
   
![image](https://github.com/slskd/slskd/assets/4962388/b63dc781-5e2c-43d7-aef2-52e77a09ab63)

**Comment by perioecus** [comment]
Similar situation for me: slskd runs in an archlinux incus container with ~280G of shared files. 
![2024-04-26-20-41-32_selected_screenshot](https://github.com/slskd/slskd/assets/39342219/08ee64aa-6062-4463-b7ee-8a47b82d728b)

**Comment by jpdillingham** [comment]
Apologies for the delayed response!

tl;dr I don't have an answer.  I don't believe slskd has any memory leaks as I constantly monitor for them, but I have not done exhaustive research into memory usage with large shares.

A while ago I did add the ability to control whether the share cache is stored in memory or on disk (docs [here](https://github.com/slskd/slskd/blob/master/docs/config.md#cache)).  This was originally designed for low spec hardware like a Raspberry Pi or other SBC, but would also apply to those that want to minimize memory usage.  You can set `storage_mode` to `disk` and you _should_ see VIRT drop at the expense of higher CPU usage.  If you'd be willing to give this a try and report back, you'd be helping me out!

I've been trying to better educate myself about virtual memory (specifically Linux based) along with the "process working set" statistic I get from .NET performance counters.

Per `man top`:

```
 The  total  amount  of  virtual  memory  used  by the task.  It
      includes all code, data and shared libraries  plus  pages  that
      have  been  swapped out and pages that have been mapped but not
      used.
```

'Process working set':

```
The working set of a process is the set of memory pages currently visible to the process in physical RAM memory. These pages are resident and available for an application to use without triggering a page fault.

The working set includes both shared and private data. The shared data includes the pages that contain all the instructions that the process executes, including the process modules and the system libraries.
```

I have a hunch that these values can include the size of files that have been opened for reading, which what slskd will do when performing a scan.

[This](https://unix.stackexchange.com/questions/479675/top-showing-virtual-memory-usage-of-hundreds-of-gigabytes) StackExchange post would seem to agree with that, though I'm not 100% sure "map" is the best way to characterize what slskd does when scanning files.

[Here](https://serverfault.com/questions/138427/what-does-virtual-memory-size-in-top-mean) is a similar post on Serverfault, which states (in part):

```
Virtual memory isn't even necessarily memory. For example, if a process memory-maps a large file, the file is actually stored on disk, but it still takes up "address space" in the process.

Address space (ie. virtual memory in the process list) doesn't cost anything; it's not real. What's real is the RSS (RES) column, which is resident memory. That's how much of your actual memory a process is occupying.
```

So, in conclusion, I'm really not sure.  But I don't _think_ it is a huge issue.

**Comment by mathiascode** [comment]
Does this still happen with slskd 0.24.1?

**Comment by jpdillingham** [comment]
I continuously monitor the canary and I haven't seen a problem.  I think this is either a reporting anomaly or it is just inherent to the .NET runtime (perhaps this is incorrect, but I can't identify anything actionable)

**Comment by jpdillingham** [comment]
I went down this rabbit hole again today and found [this](https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#region-range) in Microsoft's documentation for the garbage collector.

The dotnet runtime reserves (apparently) 5 times the amount of the GC hard limit (currently 2gb in the Dockerfile) of virtual memory at startup.  I did some investigation into the instance I run for testing:

```bash
jp@canary:~$ sudo pmap -x 1731 | sort -k2 -n -r | head -40
00007efbc1401000 10399740       0       0 -----   [ anon ]
000055ebb25d0000 1912184       0       0 -----   [ anon ]
```

We see a ~10gb reservation followed by ~2gb, and Claude makes a convincing argument that these are this the virtual memory reservation and the GC allocation, respectively.  

The docs:

> The recommendation for the reservation amount is two to five times the committed size for your GC heap. If your scenario does not make many large allocations (this could be any allocations on UOH or larger than the UOH region size), twice the committed size should be safe. Otherwise, you might want to make it larger so you don't incur too frequent full-compacting GCs to make space for those larger regions. If you don't know your GC heap's committed size, you can set this to two times the amount of physical memory available to your process.

So, if someone was not satisfied with the VIRT slskd uses, they could set `DOTNET_GCRegionRange` to:

```
DOTNET_GCRegionRange=0x100000000   # 4 GB — 2× the hard limit, on the low end of the recommended range
DOTNET_GCRegionRange=0xC0000000    # 3 GB — 1.5× the hard limit
```

Lowering this number increases the risk of crashes caused by out of memory exceptions, so users should understand the risks if they do choose to make this adjustment.
